### PR TITLE
Ported RxHttpExchangeScope to RxJava2

### PR DESCRIPTION
**Pull request by tobias-bahls**
* Wrote some unit tests around `HttpExchangeScope`
* Ported `RxHttpExchangeScope` to RxJava2.

### CODE REVIEW

File: src/main/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScope.java [hunk 1]
```diff
@@ -4,19 +4,27 @@
...
+    static <T> ObservableOperator<T, T> scoped(final HttpExchangeScope scope, final HttpServerExchange exchange) {
         requireNonNull(exchange);
 
-        return child -> {
-            return new Subscriber<T>(child) {
+        return (Observer<? super T> child) ->
+                new Observer<T>() {
+                @Override
+                public void onSubscribe(final Disposable d) {
+                    child.onSubscribe(d);
```
Comment (cberg-zalando) [inline src/main/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScope.java]: Hmm… Shouldn't we let this also be run inside `scope.runScoped`? Typically you don't do so much stuff in `onSubscribe`, but you never know what clients tend to do.

File: src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
+package org.zalando.undertaking.inject.guice;
+
+import static org.hamcrest.Matchers.*;
+
+import static org.junit.Assert.assertThat;
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java]: Please use the `assertThat` import of Hamcrest instead.

File: src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        Injector injector = getInjector();
+        HttpExchangeScope scope = injector.getInstance(HttpExchangeScope.class);
+        HttpServerExchange demoExchange = new HttpServerExchange(null);
+
+        demoExchange.getRequestHeaders().put(HttpString.tryFromString("X-Some-Header"), "blah");
+
+        scope.scoped(exchange -> {
+                 TestSimpleHandler demoHandler = getNamedHandler(TestSimpleHandler.class, "testhandler", injector);
+
+                 assertThat(demoHandler.getHeaderMap(), is(not(nullValue())));
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java]: In this case, the `is` is superfluous.

File: src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                 assertThat(demoHandler.getHeaderMap().getFirst("X-Some-Header"), equalTo("blah"));
+             }).handleRequest(demoExchange1);
+
+        HttpServerExchange demoExchange2 = new HttpServerExchange(null);
+        demoExchange2.getRequestHeaders().put(HttpString.tryFromString("X-Some-Header"), "potter");
+        scope.scoped(exchange -> {
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java]: You should add a newline here or remove it to look the same as the first block.

File: src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+             }).handleRequest(demoExchange);
+    }
+
+    @Test
+    public void injectsRequestScopedHeaderMapOverTwoInjections() throws Exception {
+        Injector injector = getInjector();
+        HttpExchangeScope scope = injector.getInstance(HttpExchangeScope.class);
+
+        HttpServerExchange demoExchange1 = new HttpServerExchange(null);
+        demoExchange1.getRequestHeaders().put(HttpString.tryFromString("X-Some-Header"), "blah");
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java]: Shouldn't this be `harry` then? :smirk:

File: src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        scope.scoped(exchange -> {
+                 TestAllInjectedHandler demoHandler = getNamedHandler(TestAllInjectedHandler.class,
+                         "testAllInjectedHandler", injector);
+
+                 assertThat(demoHandler.getExchangeScopedExecutor(), is(not(nullValue())));
+             }).handleRequest(demoExchange);
+    }
+
+    @Test
+    public void throwsIfHttpExchangeScopedIsUsedOutOfScope() {
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/guice/HttpExchangeScopeModuleTest.java]: Perhaps a little documentation might be good. Something like:

```
/**
 * Verifies that object explicitly bound in the `HttpExchangeScope` can not be accessed outside.
 *
 * @see #getInjector()
 */
```

File: src/test/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScopeTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    private Injector injector;
+
+    @Before
+    public void setUp() throws Exception {
+        injector = getInjector();
+    }
+
+    @Test
+    public void throwsWithoutScopedExcutionObservable() {
+        Injector injector = getInjector();
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScopeTest.java]: Why do you need to request the injector again and not use the one you already assigned to the class attribute?

File: src/test/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScopeTest.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        demoExchange.getRequestHeaders().put(HttpString.tryFromString("X-Some-Header"), "blah");
+
+        Observable.just(1)                                                         //
+                  .lift(scope.scoped(demoExchange))                                //
+                  .flatMap((e) -> getSimpleHandlerSingle(injector).toObservable()) //
+                  .map(TestSimpleHandler::getHeaderMap)                            //
+                  .test()                                                          //
+                  .assertComplete()                                                //
+                  .assertValue(headers ->                                          //
+                           headers.getFirst("X-Some-Header").equals("blah"));      //
```
Comment (cberg-zalando) [inline src/test/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScopeTest.java]: You do not need the `//` on the last line.

File: src/main/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScope.java (hunk 1 above)
Comment (tobias-bahls) [inline src/main/java/org/zalando/undertaking/inject/rx/RxHttpExchangeScope.java]: Good point

**Comment by cberg-zalando** [comment]
:+1:

**Comment by ehrlichandreas** [comment]
👍

**Pull request merged**